### ISSUE TITLE
Ftp Connector: Adding last modified date and file size to FtpFile

### ISSUE THREAD

**Issue by lowchinwei** [akka/alpakka#153]
It is useful to return extra file metadata such as last modified date and file size in the ls api.

**Comment by juanjovazquez** [comment]
@lowchinwei Definitely, `FTPFile` object needs to be enriched. I need to take some time out for tackling this ;).

**Comment by svezfaz** [comment]
@juanjovazquez (1) modified date, (2) file size and (3) file permissions seem like the lowest-hanging fruits here.. The resulting model object would look something like

```
final case class FtpFile(
    name: String,
    path: String,
    isDirectory: Boolean,
    size: Long,
    lastModified: java.time.LocalDateTime,
    permissions: Set[java.nio.file.attribute.PosixFilePermission]
) {...}
```

Info about user and group is not that low-hanging, as FTP will give us nice names, where SFTP will give us numerical ids only.

Would a first PR to add the 3 above be welcome?

**Comment by eharik** [comment]
Could you add a couple parameters to FtpFile so it's FtpFile[UID, GID] or
is that over complicating things?  Either way, the richer FtpFile changes
will be welcome!

On Feb 16, 2017 4:00 PM, "Stefano Bonetti" <notifications@github.com> wrote:

@juanjovazquez <https://github.com/juanjovazquez> (1) modified date, (2)
file size and (3) file permissions seem like the lowest-hanging fruits
here.. The resulting model object would look something like

final case class FtpFile(
    name: String,
    path: String,
    isDirectory: Boolean,
    size: Long,
    lastModified: java.time.LocalDateTime,
    permissions: Set[java.nio.file.attribute.PosixFilePermission]
) {...}

Info about user and group is not that low-hanging, as FTP will give us nice
names, where SFTP will give us numerical ids only.

Would a first PR to add the 3 above be welcome?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/akka/alpakka/issues/153#issuecomment-280458814>, or mute
the thread

**Comment by juanjovazquez** [comment]
@svezfaz the PR would be definitely welcome. Just a couple of things. `java.time.LocalDateTime` would force us to use Java 8, right?. I think this is still a strong requirement. What about the humble `java.util.Date`? (I know this API is not reliable but it could be a sensible decision). I'd rather go with joda time but I'm not sure if it's worth it in this case 'cause maybe the end user is not comfortable with the idea of including this additional dependency. OTOH, the `isDirectory` attribute is not helping much now as the `Source` is only generating proper file elements, i.e. `isDirectory` will always be `false`. I think we should change this in the context of other issue and the `Source` should generate both files and directories and let the user decide how to deal with that. Opinions?.

BTW, @eharik do you have a use case that justifies the `uid` and `gid` attributes?.

**Comment by svezfaz** [comment]
@juanjovazquez @eharik 
- point taken on `LocalDateTime`. I am not super-keen on `util.Date` but the only alternative seems to be just a Long epoch timestamp. Too low level?
- for the `isDirectory` matter, indeed it makes sense to solve this as a separate issue. so we might leave this in for now?
- for the UID/GID, maybe we can skip this for now until we find a valid use case.

**Comment by juanjovazquez** [comment]
I'd have a look at [GenericFile](https://github.com/apache/camel/blob/52a739feb9da8acd29067304c7c8356bbc5ef4dd/camel-core/src/main/java/org/apache/camel/component/file/GenericFile.java) on the Camel project. Take into account that many Alpakka users will come from Camel (I'm one of them ;-)). I can see that `lastModified` is defined as `long` there, so I'd stick to this criteria.

Agreed with the rest of considerations.

**Comment by eharik** [comment]
Nope, no compelling use case right now.  Just throwing out the suggestion.

**Comment by patriknw** [comment]
JDK8 is required anyway, by Akka

**Comment by ktoso** [comment]
Solved in #195
